server/service/apple_mdm.go

```diff
@@ -822,6 +823,141 @@ func generateEnrollmentProfileMobileconfig(orgName, fleetURL, scepChallenge, top
+}
+
+func (svc *Service) EnqueueMDMAppleCommandRemoveEnrollmentProfile(ctx context.Context, hostID uint) error {
+	if err := svc.authz.Authorize(ctx, &fleet.Host{}, fleet.ActionList); err != nil {
+		return err
+	}
+
+	h, err := svc.ds.HostLite(ctx, hostID)
+	if err != nil {
+		return ctxerr.Wrap(ctx, err, "getting host info for mdm apple remove profile command")
+	}
+
+	// check authorization again based on host info for team-based permissions
+	if err := svc.authz.Authorize(ctx, h, fleet.ActionCommand); err != nil {
+		return err
+	}
+
+	enabled, err := svc.ds.GetNanoMDMEnrollmentStatus(ctx, h.UUID)
+	switch {
+	case err != nil:
+		return ctxerr.Wrap(ctx, err, "getting mdm enrollment status")
+	case !enabled:
+		return fleet.NewUserMessageError(ctxerr.New(ctx, fmt.Sprintf("mdm is not enabled for host %d", hostID)), http.StatusConflict)
+	default:
+		// ok
+	}
+
+	cmdUUID, err := svc.enqueueMDMAppleCommandRemoveEnrollmentProfile(ctx, h.UUID)
+	if err != nil {
+		return ctxerr.Wrap(ctx, err, "enqueuing mdm apple remove profile command")
+	}
+
+	return svc.pollResultMDMAppleCommandRemoveEnrollmentProfile(ctx, cmdUUID, h.UUID)
+}
+
+func (svc *Service) enqueueMDMAppleCommandRemoveEnrollmentProfile(ctx context.Context, hostUUID string) (string, error) {
+	cmd := new(mdm.Command)
+	cmdUUID := uuid.New().String()
+	cmd.CommandUUID = cmdUUID
+	cmd.Command.RequestType = "RemoveProfile"
+	cmd.Raw = []byte(generateMDMAppleCommandRemoveEnrollmentProfile(cmdUUID, apple_mdm.FleetPayloadIdentifier))
+
+	status, res, err := rawCommandEnqueue(ctx, svc.mdmStorage, svc.mdmPushService, cmd, []string{hostUUID}, false, svc.logger)
+	if err != nil {
+		return cmdUUID, ctxerr.Wrap(ctx, err)
+	}
+
+	// Additional logging to aid in debugging unexpected results
+	if status != http.StatusOK {
+		logs := []interface{}{
+			"msg", fmt.Sprintf("enqueuing mdm apple remove profile command resulted in unexpected status %d", status),
+			"host_uuid", hostUUID,
+			"command_uuid", cmdUUID,
+		}
+		if res != nil {
+			if r, ok := res.Status[hostUUID]; !ok {
+				logs = append(logs, "detail", "none")
+			} else {
+				logs = append(logs, "push_result", r.PushResult, "push_error", r.PushError, "command_error", r.CommandError)
+			}
+		}
+		level.Debug(svc.logger).Log(logs...)
+	}
+
+	return cmdUUID, nil
+}
+
+func generateMDMAppleCommandRemoveEnrollmentProfile(cmdUUID string, profileUUID string) string {
+	return fmt.Sprintf(`
+<?xml version="1.0" encoding="UTF-8"?>
+<!DOCTYPE plist PUBLIC "-//Apple//DTD PLIST 1.0//EN" "http://www.apple.com/DTDs/PropertyList-1.0.dtd"&gt;
+<plist version="1.0">
+<dict>
+	<key>CommandUUID</key>
+	<string>%s</string>
+	<key>Command</key>
+	<dict>
+		<key>RequestType</key>
+		<string>RemoveProfile</string>
+		<key>Identifier</key>
+		<string>%s</string>
+	</dict>
+</dict>
+</plist>`, cmdUUID, profileUUID)
+}
+
+func (svc *Service) pollResultMDMAppleCommandRemoveEnrollmentProfile(ctx context.Context, cmdUUID string, deviceID string) error {
```

Comment:
you won't be able to test this in an integration test until we have some kind of mock for Apple's APIs, but would it be possible to add a test with `ds` mocked?

Reply:
@gillespi314 I've just realized that we don't have mocks for the nano functions that modify the database, and adding those is not trivial due to how `mockimpl` works (or at least I haven't found an easy way)

please feel free to ignore this comment.